Gems/ScriptCanvas/Code/Include/ScriptCanvas/AutoGen/ScriptCanvasAutoGenRegistry.cpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <algorithm>
+
+#include "ScriptCanvasAutoGenRegistry.h"
+
+namespace ScriptCanvas
+{
+    AutoGenRegistry::~AutoGenRegistry()
+    {
+        m_functions.clear();
+    }
+
+    AutoGenRegistry* AutoGenRegistry::GetInstance()
+    {
+        // Use static object so each module will keep its own registry collection
+        static AutoGenRegistry s_autogenRegistry;
+
+        return &s_autogenRegistry;
+    }
+
+    void AutoGenRegistry::ReflectFunctions(AZ::ReflectContext* context)
+    {
+        if (auto registry = GetInstance())
+        {
+            for (auto& iter : registry->m_functions)
+            {
+                if (iter.second)
+                {
+                    iter.second->Reflect(context);
+                }
+            }
+        }
+    }
+
+    void AutoGenRegistry::RegisterFunction(const char* className, IScriptCanvasFunctionRegistry* registry)
+    {
+        if (m_functions.find(className) == m_functions.end() && registry != nullptr)
+        {
+            m_functions.emplace(className, registry);
+        }
```

Comment:
it may be worth adding an AZ_Warning if the function name already exists, if there is a name collision, it could lead to missing functions.
I'm thinking perhaps we can add a field to the autogen xml file to let users identify their gem or module, we can use that information to provide detailed reports for example:

_"Attempted to register function: Math_Sqrt from Gem: MyNewGem, but that function already exists in Gem: CoreMathGem"_